### PR TITLE
WIP: Futures03

### PR DESCRIPTION
**Pull request by marcelbuesing**
Based on #423

I started migrating the audio and core crates to futures 0.3, tokio 0.2, hyper 0.13 ... There is still quite a lot left to do. I think the remaining crates could then first use futures compat and then be migrated one by one.

### CODE REVIEW

File: core/src/channel.rs [hunk 1]
```diff
@@ -129,16 +133,15 @@ impl Channel {
 }
 
 impl Stream for Channel {
-    type Item = ChannelEvent;
-    type Error = ChannelError;
+    type Item = Result<Option<ChannelEvent>, ChannelError>;
```
Comment (lrbalt) [inline core/src/channel.rs]: Can this `Option` be left out since the `Future` returns `Option<Item>`?

File: audio/src/fetch.rs [hunk 2]
```diff
@@ -1,17 +1,21 @@
...
 use std::cmp::{max, min};
 use std::fs;
 use std::io::{self, Read, Seek, SeekFrom, Write};
 use std::sync::{Arc, Condvar, Mutex};
 use std::time::{Duration, Instant};
+use tokio::prelude::*;
 use tempfile::NamedTempFile;
 
-use futures::sync::mpsc::unbounded;
+use futures::channel::mpsc::unbounded;
```
Comment (leshow) [inline audio/src/fetch.rs]: Sorry just stumbled on this PR. You should use these exports from `tokio` so you don't have to worry about versioning. Same with `Stream`, `ready`,  etc. `Future` can come from `std::future::Future`

File: audio/src/fetch.rs [hunk 3]
```diff
@@ -334,36 +338,34 @@ impl AudioFileOpenStreaming {
...
             AudioFileOpen::Cached(ref mut file) => {
                 let file = file.take().unwrap();
-                Ok(Async::Ready(AudioFile::Cached(file)))
+                Poll::Ready(Ok(AudioFile::Cached(file)))
             }
         }
     }
 }
 
 impl Future for AudioFileOpenStreaming {
```
Comment (leshow) [inline audio/src/fetch.rs]: If you wanted you could probably get rid of a good deal of these manually impl'd futures.

```
async fn open(&mut self) -> Result<Self, ChannelError> {
   let (id, data) = self.headers.await?;
  ...
}
```

**Comment by ashthespy** [comment]
@marcelbuesing Nice! 

I have also a branch where I played with tokio/futures/hyper/ for a while. I had it migrated to tokio 0.1 and played with tokio compat for a while to try and do things in bits. But I guess it's time to bite the bullet and rewrite it!

**Comment by marcelbuesing** [comment]
My plan is to first migrate the `core` part and then maybe use tokio compat for the other parts. Migrating the other parts afterwards one by one. It seems like a never ending story ;) though, just migrating the core part... Do you have your attempt still somewhere accessible? I'm sure this would help me.

**Comment by ashthespy** [comment]
>My plan is to first migrate the core part and then maybe use tokio compat for the other parts. Migrating the other parts afterwards one by one. 

I initially was on the same path - migrate from `tokio-core` to `tokio 0.1` and then use `compat` to move each workspace crate.. But I quickly hit a dead end with my skills. 

This is what I have now - there might be quite some overlap..  [ashthespy/librespot/tree/tokio_03](https://github.com/ashthespy/librespot/tree/tokio_03)
be warned though, it doesn't compile, with a lot of that I need to fix..
```
 expected enum `std::task::Poll`, found enum `std::result::Result`  
error: aborting due to 58 previous errors

Some errors have detailed explanations: E0034, E0046, E0050, E0053, E0061, E0277, E0308, E0599, E0621.
For more information about an error, try `rustc --explain E0034`.
error: could not compile `librespot-core`.
```

**Comment by marcelbuesing** [comment]
Thanks @ashthespy for sharing. 

As recommended by @leshow I now use `std::task::Pin`, `std::task::Poll`, `std::future::Future`, `tokio::sync::{mpsc, oneshot}`, `tokio::stream::Stream`

I'm stuck with some parts now does anyhow have some hints?

- `tokio_io::io{ReadExact, Window, WriteAll}`none of those are exported anymore in tokio 0.2. They are exposed under futures 0.3 though. I can not use those though as tokio's `TcpStream` only implements tokio's `AsyncRead, AsyncWrite`. I know there is [tokio_util::compat](https://docs.rs/tokio-util/0.3.1/tokio_util/compat/index.html) but the forth and back between those types seems to be quite difficult to achieve.
- `Sink` should I use this from the tokio or the future crate?

**Comment by ashthespy** [comment]
So, from my understanding (which my as well be wrong!), Tokio re-exports some traits from `Futures`, but not all. Channels in tokio are different implementations though. For our useage, it would be simpler to use as much as possible from tokio, so future upgrades are easer, instead of this intertwined architecture we currently have.

 > * `tokio_io::io{ReadExact, Window, WriteAll}`none of those are exported anymore in tokio 0.2. They are exposed under futures 0.3 though. I can not use those though as tokio's `TcpStream` only implements tokio's `AsyncRead, AsyncWrite`. I know there is [tokio_util::compat](https://docs.rs/tokio-util/0.3.1/tokio_util/compat/index.html) but the forth and back between those types seems to be quite difficult to achieve.

I have been trying to get `core/connection` to work with [tokio-util/codec](https://docs.rs/tokio-util/0.3.1/tokio_util/codec/index.html). Unfortunately the docs are a bit sparse though, and I quite sure what [I've done won't work](https://github.com/ashthespy/librespot/blob/tokio_03/core/src/connection/handshake.rs) :-D. Ideally I'd like to also use this opportunity to write a few crate level tests as well, so we can work on each sub module in isolation. 

> `Sink` should I use this from the tokio or the future crate?

My naive idea is again to use as much as possible form tokio. When we hit issues with missing trait implementations, head to Futures. 

On a side note, a lot of our functions that return `Box<dyn Future<_>` can become simple  `async fn` or `impl Future` functions that can use `await` now. It does make it quite a bit more ergonomic. But then again, async flow control can be tricky, and I am a bit weary to start refactoring with no tests/sanity checks in place..

**Comment by leshow** [comment]
You should be able to import AsyncWriteExt and AsyncReadExt, which provides read_exact & friends. This is implemented for TcpStream also.

**Comment by ashthespy** [comment]
> You should be able to import AsyncWriteExt and AsyncReadExt, which provides read_exact, write_all, etc. This is implemented for TcpStream also.

@leshow Could you take a peek into  [`handshake.rs`](https://github.com/librespot-org/librespot/blob/e8ec5feb2260e9d2e3b8bfc927cc5211d626e015/core/src/connection/handshake.rs#L24-L28) to confirm that requires also a rewrite?
For example, from what I see, `AsyncRead::read_exact` doesn't expose the `ReadExact` struct correct?

**Comment by leshow** [comment]
Tokio no longer provides a `ReadExact` struct, they provide `AsyncReadExt` which has a method `read_exact` https://docs.rs/tokio/0.2.11/tokio/io/trait.AsyncReadExt.html#method.read_exact that you are intended to call in an async function. I don't think they export the `read_exact` function anymore, you're intended to implement `AsyncRead` for a type then get all the `Ext` methods for free.

`tokio_io` that you're importing is out of date and not used anymore afaik, all of the tokio family of crates has been merged into `tokio` and put behind feature gates, so to get access to `AsyncReadExt` for example, you'd add the feature `io-util`.

I don't know off the top of my head how you'd write a proper `Future` impl for this type. I'd probably just go with a regular async function and scrap it all, I'm not that great at writing `Future` impls. 

Of course, once you start that, you have to do it pretty much everywhere.

**Comment by ashthespy** [comment]
@marcelbuesing I bit the bullet and thanks for some pointers from the tokio discord, rewrote the connection/handshake stuff in async/await, I'll clean it up and push it soon..

**Comment by leshow** [comment]
I'm open to help convert some of this if you get it started

**Comment by marcelbuesing** [comment]
I'm closing this in favor of #487

**Pull request closed without merging**